### PR TITLE
276 spec info

### PR DESCRIPTION
**Pull request by araszka**

### CODE REVIEW

File: src/EvoSC.Common/Interfaces/Models/IPlayer.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ public interface IPlayer : IEquatable<IPlayer>
     /// <summary>
     /// The Player's account ID.
     /// </summary>
-    public string AccountId { get; }
+    public string  AccountId { get; }
```
Comment (snixtho) [inline src/EvoSC.Common/Interfaces/Models/IPlayer.cs:18]:
```suggestion
    public string AccountId { get; }
```

File: src/Modules/GameModeUiModule/Models/GameModeUiComponentSettings.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿namespace EvoSC.Modules.Official.GameModeUiModule.Models;
+
+public class GameModeUiComponentSettings(string name, bool visible, double x, double y, double scale)
```
Comment (snixtho) [inline src/Modules/GameModeUiModule/Models/GameModeUiComponentSettings.cs:3]: add comments

File: src/Modules/GameModeUiModule/Services/GameModeUiModuleService.cs [hunk 3]
```diff
@@ -2,148 +2,159 @@
...
 using EvoSC.Modules.Official.GameModeUiModule.Config;
+using EvoSC.Modules.Official.GameModeUiModule.Enums;
 using EvoSC.Modules.Official.GameModeUiModule.Interfaces;
+using EvoSC.Modules.Official.GameModeUiModule.Models;
 using Newtonsoft.Json;
 
 namespace EvoSC.Modules.Official.GameModeUiModule.Services;
 
-[Service(LifeStyle = ServiceLifeStyle.Transient)]
+[Service(LifeStyle = ServiceLifeStyle.Singleton)]
```
Comment (snixtho) [inline src/Modules/GameModeUiModule/Services/GameModeUiModuleService.cs:12]: Doesn't seem like it needs to be a singleton?

File: src/Modules/SpectatorCamModeModule/Config/ISpectatorCamModeSettings.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+﻿using System.ComponentModel;
+using Config.Net;
+using EvoSC.Common.Util.Manialinks;
+using EvoSC.Modules.Attributes;
+
+namespace EvoSC.Modules.Official.SpectatorCamModeModule.Config;
+
+[Settings]
+public interface ISpectatorCamModeSettings
```
Comment (snixtho) [inline src/Modules/SpectatorCamModeModule/Config/ISpectatorCamModeSettings.cs:9]: I think adding a config option to enable/disable the custom spectator switch would be useful, or move that functionality to a separate module.

File: src/Modules/SpectatorCamModeModule/Templates/SpectatorMode.mt [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+                <frameinstance modelid="option"/>
+                <frameinstance modelid="option"/>
+                <frameinstance modelid="option"/>
+                <frameinstance modelid="option"/>
+            </frame>
+        </frame>
+    </template>
+
+    <script><!--
+    #Include "TextLib" as TextLib
```
Comment (snixtho) [inline src/Modules/SpectatorCamModeModule/Templates/SpectatorMode.mt:55]: With this amount of maniascript, move it to a file to make the overall code more organized.

Also some of the code appears to be copied from somewhere? If so, it's important to add references in the comments, with links if possible.

File: src/Modules/SpectatorTargetInfoModule/Services/SpectatorTargetInfoService.cs [hunk 6]
```diff
@@ -1,99 +1,250 @@
...
+
+        if (_spectatorTargets.TryGetValue(spectatorLogin, out var target) && target == targetPlayer)
+        {
+            return null; //Player is already spectating target
+        }
+
+        _spectatorTargets[spectatorLogin] = targetPlayer;
 
-        return server.Remote.TriggerModeScriptEventArrayAsync("Common.UIModules.SetProperties", hudSettings.ToArray());
+        logger.LogDebug("Updated spectator target {spectatorLogin} -> {targetLogin}.", spectatorLogin,
```
Comment (snixtho) [inline src/Modules/SpectatorTargetInfoModule/Services/SpectatorTargetInfoService.cs:111]: Perhaps this should be trace? Might be quite verbose

File: src/Modules/SpectatorTargetInfoModule/Services/SpectatorTargetInfoService.cs [hunk 7]
```diff
@@ -1,99 +1,250 @@
...
-        };
+            await SendSpectatorInfoWidgetAsync(spectatorLogin, targetPlayer);
+        }
+    }
+
+    public Task RemovePlayerFromSpectatorsListAsync(string spectatorLogin)
+    {
+        if (_spectatorTargets.Remove(spectatorLogin))
+        {
+            logger.LogDebug("Removed spectator {spectatorLogin}.", spectatorLogin);
```
Comment (snixtho) [inline src/Modules/SpectatorTargetInfoModule/Services/SpectatorTargetInfoService.cs:130]: Again like comment above about trace.

File: src/Modules/SpectatorTargetInfoModule/Templates/SpectatorTargetInfo.mt [hunk 8]
```diff
@@ -1,280 +1,254 @@
...
+
     <script><!--
-    #Include "MathLib" as ML
-    #Include "TextLib" as TL
-        
-    #Const C_Status_Disconnected	0
-    #Const C_Status_Spawned			1
-    #Const C_Status_NotSpawned		2
-    #Const C_Status_Spectating		3
+    #Include "Libs/Nadeo/CMGame/Utils/Icons.Script.txt" as Icons
```
Comment (snixtho) [inline src/Modules/SpectatorTargetInfoModule/Templates/SpectatorTargetInfo.mt:111]: move maniascript to separate file again

File: src/Modules/TeamInfoModule/Controllers/TeamInfoEventController.cs [hunk 9]
```diff
@@ -37,6 +38,17 @@ public async Task OnScoresAsync(object sender, ScoresEventArgs eventArgs)
             await teamInfoService.SetModeIsTeamsAsync(true);
         }
 
+        var team1 = eventArgs.Teams.ToList()[0];
+        if (team1 != null)
+        {
+            logger.LogInformation("[TEAM 1] Match: {matchPoints} Map: {mapPoints}", team1.MatchPoints, team1.MapPoints);
```
Comment (snixtho) [inline src/Modules/TeamInfoModule/Controllers/TeamInfoEventController.cs:44]: debug/trace

File: src/Modules/TeamInfoModule/Controllers/TeamInfoEventController.cs [hunk 10]
```diff
@@ -37,6 +38,17 @@ public async Task OnScoresAsync(object sender, ScoresEventArgs eventArgs)
...
+        }
+        var team2 = eventArgs.Teams.ToList()[1];
+        if (team2 != null)
+        {
+            logger.LogInformation("[TEAM 2] Match: {matchPoints} Map: {mapPoints}", team2.MatchPoints, team2.MapPoints);
```
Comment (snixtho) [inline src/Modules/TeamInfoModule/Controllers/TeamInfoEventController.cs:49]: debug/trace

File: tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+        return Task.CompletedTask;
+    }
+
+    [Fact]
+    public Task Returns_Null_For_Non_Existent_Player_data()
+    {
+        var checkpointsGroup = new CheckpointsGroup();
+
+        Assert.Null(checkpointsGroup.GetPlayerCheckpointData("*fakeplayer1*"));
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs:33]: move "act" statements out of asserts

File: tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        var checkpointsGroup = new CheckpointsGroup();
+        var targetPlayer = new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer1*" };
+
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer10*" }, 0));
+        checkpointsGroup.Add(new CheckpointData(targetPlayer, 1));
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer12*" }, 2));
+
+        Assert.Equal(1, checkpointsGroup.GetPlayerCheckpointData(targetPlayer.GetLogin())?.time);
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs:23]: move "act" statements out of asserts (remember the arrange, act,  assert (AAA) structure)

File: tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        var checkpointsGroup = new CheckpointsGroup();
+        var targetPlayer = new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer1*" };
+
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer10*" }, 0));
+        checkpointsGroup.Add(new CheckpointData(targetPlayer, 1));
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer12*" }, 2));
+
+        Assert.Equal(2, checkpointsGroup.GetRank(targetPlayer.AccountId));
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs:50]: move "act" statements out of asserts

File: tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer10*" }, 0));
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer11*" }, 1));
+        checkpointsGroup.Add(new CheckpointData(targetPlayer, 1));
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer12*" }, 2));
+
+        Assert.Equal(3, checkpointsGroup.GetRank(targetPlayer.AccountId));
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs:69]: move "act" statements out of asserts

File: tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer10*" }, 0));
+        checkpointsGroup.Add(new CheckpointData(targetPlayer, 1));
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer11*" }, 1));
+        checkpointsGroup.Add(
+            new CheckpointData(new OnlinePlayer { State = PlayerState.Playing, AccountId = "*fakeplayer12*" }, 2));
+
+        Assert.Equal(2, checkpointsGroup.GetRank(targetPlayer.AccountId));
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Models/CheckpointsGroupTests.cs:88]: move "act" statements out of asserts

File: tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+        Assert.Equal(2, cpTimes.Count);
+        Assert.Single(cpTimes[1]);
+        Assert.Equal(1000, cpTimes[1][0].time);
+        Assert.Equal(3, cpTimes[2].Count);
+        Assert.Equal(1000, cpTimes[2][0].time);
+        Assert.Equal(1200, cpTimes[2][1].time);
+        Assert.Equal(1400, cpTimes[2][2].time);
+
+        await spectatorTargetService.ClearCheckpointsAsync();
+        Assert.Empty(spectatorTargetService.GetCheckpointTimes());
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs:74]: move "act" statements out of asserts

File: tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+            ),
+            new(
+                new OnlinePlayer
+                {
+                    AccountId = PlayerUtils.ConvertLoginToAccountId("*fakeplayer3*"), State = PlayerState.Spectating
+                }, 1_111
+            ),
+        };
+
+        Assert.Equal(1, checkpointsList.GetRank("*fakeplayer1*"));
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs:198]: move "act" statements out of asserts

File: tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+
+    [Theory]
+    [InlineData(900, 1_000, 100)]
+    [InlineData(100, 999, 899)]
+    [InlineData(400, 200, -200)]
+    public Task Calculates_Time_Difference(int leadingTime, int trailingTime, int expectedTime)
+    {
+        var spectatorTargetService = ServiceMock();
+
+        Assert.Equal(expectedTime, spectatorTargetService.GetTimeDifference(leadingTime, trailingTime));
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs:214]: move "act" statements out of asserts

File: tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs [hunk 19]
```diff
@@ -0,0 +1,428 @@
...
+            .ReturnsAsync(DefaultModeScriptName.Teams);
+        _server.Remote.Setup(remote => remote.GetTeamInfoAsync(1))
+            .ReturnsAsync(new TmTeamInfo { RGB = "FF0066" });
+        _server.Remote.Setup(remote => remote.GetTeamInfoAsync(2))
+            .ReturnsAsync(new TmTeamInfo { RGB = "111111" });
+
+        await spectatorTargetService.UpdateIsTeamsModeAsync();
+        await spectatorTargetService.FetchAndCacheTeamInfoAsync();
+
+        Assert.Equal("FF0066", spectatorTargetService.GetTeamColor(PlayerTeam.Team1));
```
Comment (snixtho) [inline tests/Modules/SpectatorTargetInfoModule.Tests/Services/SpectatorTargetInfoServiceTests.cs:263]: move "act" statements out of asserts

**Review by snixtho — APPROVED** [review]

File: src/Modules/GameModeUiModule/Services/GameModeUiModuleService.cs (hunk 3 above)
Comment (araszka) [reply src/Modules/GameModeUiModule/Services/GameModeUiModuleService.cs:12]: Yeah it doesnt, I had made some changes that I reverted and forgot to change it back to transient.

File: src/Modules/SpectatorCamModeModule/Config/ISpectatorCamModeSettings.cs (hunk 4 above)
Comment (araszka) [reply src/Modules/SpectatorCamModeModule/Config/ISpectatorCamModeSettings.cs:9]: wdym it is a independent module, that can be disabled?

**Pull request merged**